Sources/Datadog/Core/System/AppStateListener.swift

```diff
@@ -0,0 +1,148 @@
+        // filter changes outside of the range
+        taken.changes = taken.changes.filter { range.contains($0.date) }
+        return taken
+    }
+
+    var foregroundDuration: TimeInterval {
+        var duration: TimeInterval = 0.0
+        var lastActiveStartDate: Date?
+        let allEvents = [initialState] + changes + [finalState]
+        for event in allEvents {
+            if let startDate = lastActiveStartDate {
+                duration += event.date.timeIntervalSince(startDate)
+            }
+            if event.isActive {
+                lastActiveStartDate = event.date
+            } else {
+                lastActiveStartDate = nil
+            }
+        }
+        return duration
+    }
+
+    var didRunInBackground: Bool {
+        return !initialState.isActive || !finalState.isActive
+    }
+
+    private func isActive(at date: Date) -> Bool {
+        if date <= initialState.date {
+            // we assume there was no change before initial state
+            return initialState.isActive
+        } else if finalState.date <= date {
+            // and no change after final state
+            return finalState.isActive
+        }
+        var active = initialState.isActive
+        for change in changes {
+            if date < change.date {
+                break
+            }
+            active = change.isActive
+        }
+        return active
+    }
+}
+
+internal protocol AppStateListening: class {
+    var history: AppStateHistory { get }
+}
+
+internal class AppStateListener: AppStateListening {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    private let dateProvider: DateProvider
+    private let publisher: ValuePublisher<AppStateHistory>
+
+    var history: AppStateHistory {
+        var current = publisher.currentValue
+        current.finalDate = dateProvider.currentDate()
+        return current
+    }
+
+    private static var isAppActive: Bool {
+        return UIApplication.managedShared?.applicationState == .active
+    }
+
+    init(dateProvider: DateProvider) {
+        self.dateProvider = dateProvider
+        let currentState = Snapshot(
+            isActive: AppStateListener.isAppActive,
+            date: dateProvider.currentDate()
+        )
+        self.publisher = ValuePublisher(
+            initialValue: AppStateHistory(
+                initialState: currentState,
+                finalDate: currentState.date
+            )
+        )
+
+        let nc = NotificationCenter.default
+        nc.addObserver(self, selector: #selector(appWillResignActive), name: UIApplication.willResignActiveNotification, object: nil)
+        nc.addObserver(self, selector: #selector(appDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil)
+    }
+
+    @objc
+    private func appWillResignActive() {
+        let now = dateProvider.currentDate()
+        publisher.mutateAsync {
+            $0.add(change: Snapshot(isActive: false, date: now))
+        }
```

Comment:
Do we need to introduce `mutateAsync` and use `inout`? It seems over-optimizing it by using complex machinery (mutating value type by reference between threads). I'm not that familiar with consequences of using `inout` across threads and can't see if we synchronize it well.

We could keep it plain simple:
```swift
var value = publisher.currentValue // has negligible overhead as reads are concurrent
value.add(change: Snapshot(isActive: false, date: now))
publisher.publishAsync(value)
```
or even remove the read overhead at all by leveraging the fact that notifications are delivered on a single thread:
```swift
private lastHistory: AppStateHistory {
   didSet { publisher.publishAsync(lastHistory) }
}

private let publisher: ValuePublisher<AppStateHistory>

private func appWillResignActive() {
   var newHistory = self.lastHistory
   newHistory.add(change: Snapshot(isActive: false, date: dateProvider.currentDate()))
   self.lastHistory = newHistory
}
```